### ISSUE TITLE
mAP on VOC2007 test.

### ISSUE THREAD

**Issue by changjo** [ljanyst/ssd-tensorflow#2]
Hi,
Thanks for sharing your codes.
I modified `source_voc_pascal.py` to train **VOC2007 + VOC2012 trainval**.
And after training, I evaulated on **VOC2007 test** but the result is far below compared to the original paper's..

![ssd_infer_voc_0712_test copy](https://user-images.githubusercontent.com/4277500/34289142-65d4c212-e734-11e7-948c-94c827a62646.png)

May I have any ideas?

**Comment by ljanyst** [comment]
It probably depends on how long you trained and what percentage of data you used for validation. The result I get with the VOC12 test is 68.71% with the reference at 75.8%: http://host.robots.ox.ac.uk:8080/anonymous/WKTCBE.html

**Comment by ljanyst** [comment]
You can also try to lower the acceptance threshold or remove the cut on the number of accepted candidates. I don't particularly like the mAP metric because it very much depends on low quality candidates that you would reject anyways while doing inference.

**Comment by ljanyst** [comment]
Another interesting test would be to compare mAP of this code vs. the original implementation at the 0.5 acceptance threshold.

**Comment by changjo** [comment]
I am trying to get the mAP of the result of original paper's **PASCAL VOC2007 test**. I tried to use 8732 of default boxes and modified scales of the boxes as the paper suggested but the mAP is still about 63%.. (I have used a confidence threshold of 0.01.)
Now, I set `validation-fraction=0` and am training again.

Also I was wondering about the best overlap.
In `ssdutil.py`,

```python
def compute_overlap(box_arr, anchors_arr, threshold):
    iou = jaccard_overlap(box_arr, anchors_arr)
    overlap = iou > threshold

    good_idxs = np.nonzero(overlap)[0]
    best_idx  = np.argmax(iou)
    best = None
    good = []

    if iou[best_idx] > threshold:
        best = Score(best_idx, iou[best_idx])

    for idx in good_idxs:
        good.append(Score(idx, iou[idx]))

    return Overlap(best, good)
```

I thought, the best overlap should be always chosen...


<pre><code><s>if iou[best_idx] > threshold:</s>
best = Score(best_idx, iou[best_idx])
</code></pre>

**Comment by ljanyst** [comment]
It's not really obvious from the paper what to do with the best match and I don't remember what the caffe  code does. Since we only consider matches good if the overlap is above 0.5, it makes sense to reject the ground truth objects completely if there is no match above 0.5. Note that this penalizes very small objects and objects whose aspect ratio is much larger than 1:3.

Another thing that the caffe code does differently is the optimizer. They use SGD with step learning rate decay at certain epochs. They also double the learning rate for biases compared to the weights. You can give that a go even though it's quite painful to do in TensorFlow.

**Comment by changjo** [comment]
I should not delete the line `if iou[best_idx] > threshold:` because this condition is needed to sample patches. I found that for PASCAL VOC 2007,
> Objects marked as `difficult` (section 2.4) are currently ignored by the evaluation. [link](http://host.robots.ox.ac.uk/pascal/VOC/voc2007/htmldoc/voc.html#SECTION00054000000000000000)

I am trying to get a performance as much as original's. I haven't found a tensorflow code achieving the performance of 77.2% for a training set of `VOC2007+VOC2012 trainval` and a test set of `VOC2007 test`. Maybe somewhere there is...

**Comment by joaqo** [comment]
Hey @changjo I am having the same problem as you do. Haven't been able to find a tensorflow implementation that can reproduce the paper's performance. Please let me know if you ever do, the only thing I've found that gets paper level results is https://github.com/amdegroot/ssd.pytorch but its in pytorch :-(

**Comment by ljanyst** [comment]
Thanks for pointing me to the pytorch implementation @joaqo. I will have a look at it and try to figure out what I did wrong. As far as I can tell right now, the only differences between the reference and my code is me not using bipartite matching and the choice of optimizer.

**Comment by joaqo** [comment]
No problem, been looking at your implementation (currently training 2 instances of it) and I really liked it, I think its probably the best tf implementation of the algorithm on github atm.

I am also training the pytorch implementation. I will see how the results go, and see if there is anything I can do to improve your version, so there is an open, trully state of the art, SSD tensorflow implementation.

**Comment by joaqo** [comment]
Also, would you mind telling me what do you mean by "not using bipartite matching" ?

**Comment by changjo** [comment]
@joaqo I am also looking into <https://github.com/Hakuyume/chainer-ssd> when I have free time. Although I have no idea about chainer but it seems not difficult to understand the code.

**Comment by ljanyst** [comment]
Bipartite matching is the algorithm the original implementation seems to use to match the ground truth boxes to anchors. It's essentially a graph flow algorithm that ensures optimal matching. I used a simple algorithm that just favors larger overlap.

**Comment by changjo** [comment]
@ljanyst @joaqo Okay, I will also have a look at it. Thanks.

**Comment by joaqo** [comment]
Just trained the pytorch implementation, got 77mAP as reported on their readme, so it seems this is the real deal.

**Comment by ljanyst** [comment]
Thanks! I will try to spend some time on it tomorrow.

**Comment by changjo** [comment]
Hello,
I have tried to modified some parts of codes related to transforms, build_sampler, and the optimizer to use momentum but I got 74.5mAP still under 77.2mAP.
I was wondering if pre-trained VGG16 weights have effects on it and am trying to convert auther's caffe weights to tensorflow and use them.

**Comment by ljanyst** [comment]
Yeah, they might. I have made the changes to the transformers too and will be committing them today. This is what I get with the large model: http://host.robots.ox.ac.uk:8080/anonymous/2AHRD9.html The small one should finish training in a couple hours. The PyTorch implementation is getting 0.73 in this test.

**Comment by ljanyst** [comment]
When I used the same configuration of samples as PyTorch and run the trained model on the training data, I got 0.82, but the result on the test data was below 0.7.

**Comment by changjo** [comment]
For me, training **VOC07+12 trainval** and testing **VOC07 test**, I got 77.7mAP on the pytorch implementation...

**Comment by ljanyst** [comment]
I have committed a bunch of changes. The key problem was that the sampler oftentimes returned samples with no matching anchors because the algorithm produced images that were too small. These samples were still run through the model but did not contribute to the loss and were, therefore, useless. The fix for that is here: https://github.com/ljanyst/ssd-tensorflow/commit/c292e6c3624165b31c2fd4207f7e4e6958f1f734

I have also now used the exact same transforms that the original implementation does. They reflect what the original code actually does, not what the paper says. See: https://github.com/ljanyst/ssd-tensorflow/commit/57e3f0f2f07e35002b295882d0a63410f1e079f8

Furthermore, I now use the exact same anchor configuration that the original implementation does. See: https://github.com/ljanyst/ssd-tensorflow/commit/46370744b5f22c35b34842f158749f989b49cc62

It would be great if you could give these mods a try.

**Comment by ljanyst** [comment]
Ohh, I never actually run this particular test.

Could you please share the mods you made to the optimizer?

Also note that the pytorch implementation skips the difficult samples. Neither I nor the original implementation does that.

**Comment by changjo** [comment]
Yes. When evaluating APs, the pytorch implementation skips the difficult samples. For running pytorch, I have not modified codes and got the result. On your codes, I just tried to use momentum optimizer and change adjusting learning rate just like the pytorch codes.

**Comment by ljanyst** [comment]
Ohh, thanks. I did that too, but things were actually slightly worse. Could you please try what's in the git head right now?

**Comment by changjo** [comment]
Sure, I can but I run the codes about 3 days ago and in the meantime, codes are not modified...

**Comment by ljanyst** [comment]
I made these mods a month ago, just pushed them now, so they appear to be made at the beginning of February.

**Comment by changjo** [comment]
Oh sorry, I miss understood. I thought the codes of the pytorch repo (https://github.com/amdegroot/ssd.pytorch).

**Comment by changjo** [comment]
Okay. I will try to run your code soon within a few days.

**Comment by ljanyst** [comment]
Cool! Let me know what the results are! :)

**Comment by changjo** [comment]
Sure! :)

**Comment by changjo** [comment]
Hi,
I trained a model for your recent codes using **VOC2007 trainval** + **VOC2012 trainval** as train set and **VOC2007 test** as test set. I got **65.2mAP** without difficult objects and slightly below with difficult objects. By the way, I found that samples should be built with respect to `'ImageSets/Main/trainval.txt'` or `'test.txt'` not just all `*.xml` files. I just checked this...

**Comment by joaqo** [comment]
> Hello,
I have tried to modified some parts of codes related to transforms, build_sampler, and the optimizer to use momentum but I got 74.5mAP still under 77.2mAP.
I was wondering if pre-trained VGG16 weights have effects on it and am trying to convert auther's caffe weights to tensorflow and use them.

hey @changjo do you mind being more specific with what those changes entail? Maybe a PR? Cause it seems that 74.5mAP you got there is the best SSD mAP on a Tensorlfow implementation atm!

**Comment by ljanyst** [comment]
@changjo just used the momentum optimizer the same way the original paper and the pytorch implementation do. I am running this configuration now with the other changes to sample selection.

**Comment by changjo** [comment]
Hello, I have mimicked build_sampler, SamplerTransform, LabelCreatorTransform, and SamplePickerTransform of the chainer implemention and I used momentum optimizer. My codes are not clean. Maybe I need to clean them later...

**Comment by changjo** [comment]
@ljanyst Cool! I hope it produce good results.

**Comment by ljanyst** [comment]
The thing gets to be quite good with recent changes. It's still not as good as the reference, but it seems to be pretty close to the pytorch implementation now. At least in the VOC12 test: [SSD512](http://host.robots.ox.ac.uk:8080/anonymous/CNQPDK.html), [SSD300](http://host.robots.ox.ac.uk:8080/anonymous/C3DN60.html).

**Comment by changjo** [comment]
Cool! I will try to check it.

**Comment by changjo** [comment]
I got **73.2mAP** on your recent codes for **VOC2007 trainval** + **VOC2012 trainval** as train set and **VOC2007 test** as test set without `difficult` boxes.

**Comment by guillaume-guerdoux** [comment]
Hello,
I've used you model and trained it (without making any changes to your code) for 185 epochs. But my mAP wasn't improving at the end and was very far from what is anounced in the README.
I have a 0.59 mAP on training set and 0.52 ans val set (I used as you did pascal voc 2012 and 2007) 
![image](https://user-images.githubusercontent.com/25098729/38143617-9fb618a0-3441-11e8-8f81-01b73e082094.png)
![image](https://user-images.githubusercontent.com/25098729/38143625-ac1ece70-3441-11e8-836b-09739ea97004.png)

Do you know what could I've done wrong please ? 
Then, could you explain me why there is a big jump at the 120th epoch please ? Because of the learning  rate ? 

Thank you very much !

**Comment by guillaume-guerdoux** [comment]
I also run the infer.py on PascalVoc2007 trainval set (based on the trainval.txt file) and only get 0.68 mAP
![image](https://user-images.githubusercontent.com/25098729/38144069-a01c358e-3443-11e8-8104-cbe981cb5a73.png)

I'm trying to figure out what's going wrong but can't find it for the moment ^^
Thank you :)

**Comment by guillaume-guerdoux** [comment]
Hello, I infer my model by using a confidence threshold equal to 0.5 but it seems that pytorch implementation uses a threshold equal to 0.01. 
Which threshold do you use to get you mAP score ? A 0.5 or a 0.01 threshold ? :) 
Thank you !

**Comment by ljanyst** [comment]
@guillaume-guerdoux Regarding your first question. The result shown in TensorBoard is taking into account predictions at 0.5 confidence and above. It's done this way for performance reasons. Taking everything at and above 0.01 and doing full maximum suppression takes way too much time to do at every iteration.

The learning rate drops from 0.001 to 0.0001 at 320000 iterations, that's roughly 120 epochs.

**Comment by ljanyst** [comment]
@guillaume-guerdoux Regarding your second question. I trained my model for 200 epochs.

**Comment by ljanyst** [comment]
@guillaume-guerdoux Regarding your third question. I use the same confidence threshold as the original paper describes - 0.01. It's done this way to increase recall. My blog post about SSD discusses it in more details.

**Comment by guillaume-guerdoux** [comment]
@ljanyst  Thank you very much for you answers ! I didn't get the point that a 0.01 threshold was used in the original paper to process mAP (that's why I was so surprised that my results were so differents from the results you got ) :+1: 

By curiosity, why do you use now a momentum instead of the adaptive learning rate (adam) you used before ? 
Thanks

**Comment by twangnh** [comment]
Hi! guillaume-guerdoux, your validation mAP is around 0.25 at 10epoch, mine is only 0.03 at 10epoch

**Comment by ljanyst** [comment]
This is a stochastic process starting from a random state. It seems that the optimizer got stuck. Just restart it.

**Comment by ljanyst** [comment]
@guillaume-guerdoux The momentum optimizer gets better final mAP comparing to Adam, even though Adam pushes the loss much lower. This is likely the effect of L2 normalization (weight decay in Caffe parlance).

**Comment by twangnh** [comment]
compared to @guillaume-guerdoux ,I got rather poor mAP:
![training_map](https://user-images.githubusercontent.com/18298163/38463009-7c459d36-3b22-11e8-9280-6fcddf198b50.png)
![validation_map](https://user-images.githubusercontent.com/18298163/38463015-9f6771cc-3b22-11e8-9b6f-6d8c5c125c8e.png)
would the random initialization cause this issue? seems that the initialization is not stable

**Comment by twangnh** [comment]
turns out it was caused by python version

**Comment by guillaume-guerdoux** [comment]
@ljanyst  Thank you for your answer ! 
Did you try your model by adding the coco dataset  to get similar results as in the paper ? 
I tried to add it but didn't get concluding results for the moment. 
The classes are so unbalanced in coco (more than 200 000 persons while the median is around 30 000) 
What would be your tips to it ? Reduce the personn dataset ? :)

**Comment by HiKapok** [comment]
hi everyone, there are quite a lot ssd implementations in TensorFlow, but only a few of them can be trained to state-of-the-art performance, here is one another repo with SOTA performance, see [here](https://github.com/HiKapok/SSD.TensorFlow) if you are interested in. And any contribution will be welcomed.

**Comment by RoseLii** [comment]
Please help me, when I run ./export_model.py
usage: export_model.py [-h] [--metagraph-file METAGRAPH_FILE]
[--checkpoint-file CHECKPOINT_FILE]
[--output-file OUTPUT_FILE] --output-tensors
OUTPUT_TENSORS [OUTPUT_TENSORS ...]
export_model.py: error: the following arguments are required: --output-tensors
But when I run ./infer.py, test-output doesn't show anything。
when i run ./infer.py:
[i] Project name:       test
[i] Training data:      pascal-voc/training-data.pkl
[i] Batch size:         32
[i] Data source:        None
[i] Data directory:     pascal-voc
[i] Output directory:   test-output
[i] Annotate:           False
[i] Dump predictions:   False
[i] Sample:             test
[i] Threshold:          0.5
[i] Pascal summary:     False
[i] Compute stats:      False
[i] Network checkpoint: test/final.ckpt
[i] Metagraph file:     test/final.ckpt.meta
[i] Image size:         Size(w=300, h=300)
[i] Number of files:    4952
[i] Creating the model...
[i] Processing samples: 100%|██████████████████████████████████████████████████████████| 155/155 [00:54<00:00, 2.82batches/s]
[i] All done.
who can guide me? The specific training detection process, how to do your own data set. I will be very grateful. @changjo @guillaume-guerdoux @ljanyst @joaqo 
 Can you share the code you have changed? thank you very much。1103776854@qq.com

**Comment by changjo** [comment]
@HiKapok Thanks for sharing.

**Comment by szt1213** [comment]
Hello, thank you very much for sharing.I've also exported my model.But what I want to know is, can only generate .pb file, cannot produce .pbtxt file at the same time.

**Comment by Houliuyu** [comment]
@MrWanter Hello,I also got the same poor mAP,could you please tell me what is wrong with the python version?Do I need to use python3?

**Comment by ljanyst** [comment]
@changjo After the recent fix, the training now reaches the mAP described in the paper.
Others: Yes, it's a python3 program.

**Comment by changjo** [comment]
@ljanyst Oh, Thank you very much.
